editor/src/messages/tool/common_functionality/shape_editor.rs

```diff
@@ -206,6 +206,114 @@ impl ShapeEditor {
 		}
 	}
 
+	/// The opposing handle lengths.
+	pub fn opposing_handle_lengths(&self, document: &Document) -> HashMap<Vec<LayerId>, HashMap<u64, f64>> {
+		self.selected_layers()
+			.iter()
+			.filter_map(|path| document.layer(path).ok().map(|layer| (path, layer)))
+			.filter_map(|(path, shape)| shape.as_subpath().map(|subpath| (path, subpath)))
+			.map(|(path, shape)| {
+				let opposing_handle_lengths = shape
+					.manipulator_groups()
+					.enumerate()
+					.filter_map(|(id, manipulator_group)| {
+						// We will keep track of the opposing handle length when:
+						// i) Both handles exist and exactly one is selected.
+						// ii) The anchor is not selected.
+						// iii) We have to mirror the angle between handles.
+
+						if !manipulator_group.editor_state.mirror_angle_between_handles {
+							return None;
+						}
+
+						let mut selected_handles = manipulator_group.selected_handles();
+						let handle = selected_handles.next()?;
+
+						// Check that handle is the only selected handle.
+						if selected_handles.next().is_none() {
+							let opposing_handle_position = manipulator_group.opposing_handle(handle)?.position;
+							let anchor = manipulator_group.points[ManipulatorType::Anchor].as_ref()?;
+							if !anchor.is_selected() {
+								let opposing_handle_length = opposing_handle_position.distance(anchor.position);
+								Some((*id, opposing_handle_length))
+							} else {
+								None
+							}
+						} else {
+							None
+						}
+					})
+					.collect::<HashMap<_, _>>();
+				(path.clone(), opposing_handle_lengths)
+			})
+			.collect::<HashMap<_, _>>()
+	}
+
+	/// Reset the opposing handle lengths.
+	pub fn reset_opposing_handle_lengths(&self, document: &Document, opposing_handle_lengths: &HashMap<Vec<LayerId>, HashMap<u64, f64>>, responses: &mut VecDeque<Message>) {
+		self.selected_layers()
+			.iter()
+			.filter_map(|path| document.layer(path).ok().map(|layer| (path, layer)))
+			.filter_map(|(path, shape)| shape.as_subpath().map(|subpath| (path, subpath)))
+			.filter_map(|(path, shape)| opposing_handle_lengths.get(path).map(|layer_opposing_handle_lengths| (path, shape, layer_opposing_handle_lengths)))
+			.flat_map(|(path, shape, layer_opposing_handle_lengths)| {
+				shape
+					.manipulator_groups()
+					.enumerate()
+					.map(move |(id, manipulator_group)| (path, layer_opposing_handle_lengths, id, manipulator_group))
+			})
+			.for_each(|(path, layer_opposing_handle_lengths, id, manipulator_group)| {
+				if !manipulator_group.editor_state.mirror_angle_between_handles {
```

Comment:
So I've added checks to recheck whether we have to mirror angles and whether the anchor is selected - should I keep them?

Reply:
As mentioned in the `opposing_handle_lengths` function, we track the opposing length when certain conditions are met - like `mirror_angle_between_handles` being true.

The reason I had the check was because I was unsure if these properties could change without updating `opposing_handle_lengths` in `tool_data`. These might be unnecessary.

Reply:
This check probably shouldn't fail but it is fine to leave it in anyway just to make sure.